### PR TITLE
fix(hashjoin): broadcast const vectors in spill paths

### LINKED ISSUE #26432: [Bug]: unify join typed-key canonicalization across resident and spill paths
## Context

Follow-up from #26408 and #26412. The ownership/lifecycle fixes in #26412 intentionally do not redesign join-key encoding.

## Problem

Resident hash joins, initial spill partitioning, and re-spill must use one SQL-equality canonicalization contract. Today the paths can consistently disagree with SQL equality or with a nested-loop oracle for values such as:

- `DOUBLE` `+0` versus `-0`;
- NaN values;
- JSON numeric forms such as `1` versus `1.0`;
- scaled floating-point values;
- long `VECF32` values containing `+0`/`-0`.

Fixing isolated hash functions per type is insufficient: encoding, hashing, equality, partitioning, and rebuild must share the same typed-key contract.

## Proposed direction

- Define one typed-key codec/canonicalization interface used by resident build/probe, initial spill, re-spill, and remote execution.
- Specify NULL, NaN, signed-zero, JSON numeric, vector, decimal/scale, and collation behavior explicitly.
- Version any persisted spill representation affected by the codec.
- Use nested-loop execution as the semantic oracle rather than comparing two hash implementations.

## Acceptance criteria

- Resident, initial-spill, and re-spill joins return identical rows for the same inputs.
- Differential tests compare hash join variants against a nested-loop oracle for all supported key types and the edge cases above.
- Tests cover HashJoin, DedupJoin, and RightDedupJoin, including multi-column and NULL-containing keys.
- Remote/local execution and mixed spill rounds use the same codec version.
- Benchmarks show no material regression for common integer/string join keys.

Related: #26408, #26412.

### PR DESCRIPTION
**Pull request by XuPeng-SH**
## What changed

- accept non-empty constant vectors as logical broadcasts in spill and re-spill row-shape validation
- project constant varlena recovery memory from logical row ranges instead of physical vector length
- keep nil, empty constants with non-empty input, and ordinary short vectors fail-closed
- add prepared-parameter initial-spill, re-spill materialization, recovery-offset, and shape-control regressions

## Why

The typed-key work in #26738 made resident and spill hashing share one codec. After #26836, resident hash maps also accept a physical-length-one constant vector over an arbitrary logical row range, while spill validation still required `vec.Length() == rows`. A prepared parameter therefore worked in resident mode but failed initial spill or re-spill with `ErrHashBuildBudgetInvalid`.

This change aligns the existing spill boundary with the vector broadcast contract. It does not change typed-key encoding or add a new abstraction.

## Validation

The new focused tests failed on the unfixed `main` with `invalid hash build budget` and pass after the change.

- `go list`, `go build`, and `go vet` for `keycodec`, `hashbuild`, and `spillutil`
- full tests for all three owning packages
- focused HashJoin, DedupJoin, and RightDedupJoin typed-key contract suites across resident, initial spill, and re-spill modes

## Performance

The ordinary spill path adds one inlinable `IsConst` branch per vector per batch. There are no new per-row operations, allocations, copies, logs, locks, or retained state.

Closes #26432

### CODE REVIEW

File: pkg/common/hashmap/keycodec/keycodec.go [hunk 1]
```diff
@@ -27,15 +27,24 @@ import (
...
 func ValidVectors(vecs []*vector.Vector, rows int) bool {
 	if rows < 0 || len(vecs) == 0 {
 		return false
 	}
 	for _, vec := range vecs {
-		if vec == nil || vec.Length() != rows {
+		if vec == nil {
+			return false
+		}
+		if vec.IsConst() {
```
Comment (XuPeng-SH) [inline pkg/common/hashmap/keycodec/keycodec.go:42]: [P1] Preserve broadcast constants through pressure chunking

This makes a non-empty length-one const valid for any logical `rows`, but the pressure retry wrapper still slices every key with the logical `[start,end)` range. After `scatterBatchBounded` returns a retryable capacity error, `scatterBatchWithPressure` halves an 8,192-row input to 4,096 and calls `key.WindowWithAllocation(0, 4096)`; `Vector.window` rejects that because the const's physical `Length()` is 1. I reproduced this deterministically by changing the existing 80 KiB `TestSpillAllocationAccountScatterReducesUnpublishedInput` setup to pass `NewConstFixed(..., 1)` as `keyVecs`: this head returns `invalid input: invalid vector window` instead of spilling, while `JoinSpillInputReductions` is the intended recovery path. Production re-spill reaches this exact path through `scatterEvaluatedBatchWithPressure -> scatterBatchWithPressure`, so prepared-parameter re-spill still fails whenever hash/row-id allocation needs logical input reduction. The new 1 MiB direct `scatterBatchBounded` test cannot exercise it. Please make the pressure-window ownership logic const-aware (or re-evaluate/preserve the broadcast key for each logical window) and add a regression that proves a positive input-reduction count while conserving all rows. The same physical/logical distinction should be checked for const vectors in the source batch window as well.

**Review by XuPeng-SH — COMMENTED** [review]
Conclusion: **REQUEST_CHANGES**. The PR author and authenticated reviewer are both `XuPeng-SH`, so GitHub cannot record a formal self-request-changes review; this is submitted as COMMENTED.

One merge-blocking correctness gap remains. The single bounded scatter, canonical hashing, `UnionInt32` materialization, and recovery projection are otherwise consistent. Q1/Q2/Q3 found no leak, wait-cycle, or unbounded-growth defect, and the ordinary hot-path cost is one vector-level `IsConst` branch.

Validation on exact head `684797aff5f9` vs base/merge-base `0989f741286c`: all changed tests and the three full owning packages passed; HashJoin, DedupJoin, and RightDedupJoin typed-key suites passed across resident/initial-spill/re-spill. A deterministic review-only pressure regression (8,192 rows, 80 KiB account, length-one const key) fails on this head with `invalid input: invalid vector window` after input reduction, proving the inline finding. Build, vet, and `git diff --check` passed. Race stress is not applicable because the diff changes no shared/concurrent state.

**Review by LeftHandCold — APPROVED** [review]
Reviewed at 684797aff5f940a32ca84a43be670e77db87ca91. No blocking findings.

Checked the constant-vector contract end to end: non-empty constants broadcast across the batch logical row count in initial spill and re-spill, while nil vectors, empty constants for non-empty input, and ordinary short vectors remain fail-closed. The varlena recovery projection now uses logical ranges for constants without weakening non-const bounds or overflow checks. No new ownership, cleanup, retry, concurrency, or per-row hot-path work is introduced.

Local low-concurrency validation passed:
- go build and go vet for keycodec, hashbuild, and spillutil
- full tests for all three owning packages
- focused prepared-parameter/broadcast/recovery tests
- HashJoin finite-budget initial-spill + re-spill integration test

Current GitHub UT, BVT, SCA, CodeQL, and coverage checks are green.

File: pkg/common/hashmap/keycodec/keycodec.go (hunk 1 above)
Comment (XuPeng-SH) [reply pkg/common/hashmap/keycodec/keycodec.go:42]: Resolved in b95ab95d77.

The fix establishes an explicit logical-row window contract instead of special-casing this call site: scalar const values can broadcast through Batch and accounted pressure windows, while standalone Vector.Window remains physically bounded. Empty constants and short constants with row-specific grouping/provenance are rejected consistently by resident hashmap and spill validators. Offset const-null semantics and allocation ownership are preserved.

Regression coverage now forces both HashBuildSpillInputReductions and JoinSpillInputReductions with an 8K-row prepared/scalar const composite key, checks start > 0 chunks and row conservation, and covers ordinary/accounted Batch windows plus invalid metadata shapes. Full ./pkg/container/..., ./pkg/common/hashmap/..., and ./pkg/sql/colexec/... tests and go vet passed.

**Review by aptend — CHANGES_REQUESTED** [review]
## Codex deep review

### P1 — Treat broadcast const-NULL varlena as zero payload before reading row 0

`unionBatchAreaProjection` handles every non-empty const varlena with a non-empty area by calling `selectedVarlenaPayload(src, 0, 1)`. A valid const-NULL vector does not necessarily have a descriptor: `SetConstNull` clears `data` while retaining reusable `area`, and the new logical-window path also sets `data` to nil for an offset const-NULL window while continuing to borrow its area. In that state `selectedVarlenaPayload` sees zero descriptors and returns `ErrHashBuildBudgetInvalid`, although the logical payload is simply zero bytes.

This reaches the query path because shuffle HashBuild calls `projectRetainedRecovery` for every retained build batch. I reproduced it on this exact head with a deterministic counterexample: create a long varlena const of physical length one, mark row zero NULL, put it in a four-row batch, take logical window `[2,4)`, and call `projectRetainedRecovery`. The window is correctly `IsConstNull()` and still has the borrowed area, but recovery projection fails with `invalid hash build budget`. Thus a valid broadcast NULL batch can still fail under the spill/recovery path this PR is intended to close.

Please return a zero area projection for `src.IsConstNull()` before inspecting row-zero varlena data, and add the logical-window/recovery regression above. Keep the non-NULL empty/short-vector validation unchanged.

The remaining broadcast/window ownership paths close correctly: borrowed data/area have one owner, partial-window cleanup releases accounted metadata, pressure retries are bounded, and no new unbounded accumulation or wait edge was introduced.

**Comment by XuPeng-SH** [comment]
Addressed the latest request-changes review in c47df29237.

Reflection: the previous regression stopped after proving that an offset window remained `IsConstNull()`; it did not close the next consumer arc into retained-recovery projection. A const-NULL may have no row-zero descriptor while still retaining reusable varlena area, so area presence is not evidence of payload.

The production fix keeps the existing empty-const and ordinary-short-vector guards, then returns zero physical/selected area for `IsConstNull()` before descriptor inspection. It adds independent coverage for (1) offset logical-window recovery, (2) direct `SetConstNull` with stale reusable area, (3) empty-const and short-flat rejection controls, and (4) the full shuffle HashBuild lifecycle through JoinMap publication and cleanup.

Fresh validation passed: controlled focused regressions, full `./pkg/container/...`, `./pkg/common/hashmap/...`, and `./pkg/sql/colexec/...` tests, plus go list/build/vet. The hot-path change is one O(1) branch with no allocation or row scan.

@aptend please re-review.

**Review by XuPeng-SH — COMMENTED** [review]
Conclusion: APPROVE. GitHub cannot record a formal self-approval because the PR author and authenticated reviewer are both XuPeng-SH, so this verdict is submitted as COMMENTED.

I independently re-reviewed exact head c47df29237d14b93602a92ccf9f1ee45fb2113d2 against base and merge-base 00cb7ef3169a6c61f358774ac5c690a405ac162d. No blocking findings remain.

Correctness closure: scalar constants broadcast only when non-empty and free of row-specific grouping or heterogeneous provenance; empty constants, ordinary short vectors, nil vectors, and short constants with row metadata remain fail-closed. The same logical-row contract now reaches resident hashmap validation, Batch and allocation-accounted pressure windows, initial spill, re-spill, and retained-recovery projection. Const-NULL varlena contributes zero payload before descriptor inspection, including offset logical windows and SetConstNull vectors with reusable stale area.

Q1/Q2/Q3 closure: window data and area remain borrowed while owned bitmap/provenance allocations are released exactly once; partial construction and every retry return path clean their temporary windows. The pressure loop has cancellation, progress checks, bounded retries, and monotonically shrinking chunks. No new wait edge, goroutine, retained unbounded collection, or generation-crossing state is introduced.

Performance and compatibility: ordinary paths add only vector-level O(1) const/range checks. There is no new per-row work, allocation, copy, lock, logging, or persisted representation change. Standalone Vector.Window remains physically bounded; logical broadcast is exposed only through explicit logical-row and Batch APIs.

Fresh local validation on this exact head passed: focused broadcast/offset/const-NULL/stale-area/rejection/row-conservation regressions; full pkg/container/..., pkg/common/hashmap/..., and pkg/sql/colexec/... tests; controlled build and vet for all changed owners plus HashJoin, DedupJoin, and RightDedupJoin dependents; git diff --check. Race stress is not applicable because the production diff changes no concurrent/shared state.

**Review by XuPeng-SH — COMMENTED** [review]
Conclusion: APPROVE. GitHub cannot record a formal self-approval because the PR author and authenticated reviewer are both XuPeng-SH, so this verdict is submitted as COMMENTED.

I re-reviewed exact head c47df29237d14b93602a92ccf9f1ee45fb2113d2 against merge-base 00cb7ef3169a6c61f358774ac5c690a405ac162d and read every current PR comment/review. No blocking findings remain.

Both historical blockers are closed at the consumer boundary, not only at validation. Pressure reduction now uses logical-row windows for the input batch and separately evaluated key vectors, including nonzero offsets, while preserving all rows exactly once. Const-NULL varlena recovery now returns zero payload before descriptor inspection, so both offset logical windows and SetConstNull vectors with reusable stale area are valid. Empty constants for nonempty input, nil vectors, ordinary short vectors, grouping constants without the requested row domain, and heterogeneous prepared-parameter sidecars remain fail-closed.

Q1/Q2/Q3 audit: data and area in windows stay borrowed; owned bitmap and provenance allocations are released exactly once on success, partial construction failure, and every retry exit. Retry chunks shrink monotonically, cancellation and progress guards bound failure, and no wait edge, goroutine, generation crossing, or unbounded retained collection was added. Initial spill, re-spill, selected-row materialization, canonical hashing, and retained recovery now share the same logical-row contract.

Compatibility and performance: standalone Vector.Window keeps strict physical bounds; broadcast behavior is limited to explicit logical-row APIs and Batch.Window, whose RowCount owns the logical domain. Ordinary inputs add only O(vectors) constant/range checks at batch boundaries. There is no new per-row work, allocation, copy, lock, logging, wire-format change, or persisted-format change.

Fresh validation on this exact head passed: go list of the complete CGo closure; build and vet for all changed owning packages plus HashJoin, DedupJoin, RightDedupJoin, and LoopJoin consumers; full tests for those ten packages; 20 consecutive runs of the pressure row-conservation, prepared-parameter spill, offset const-NULL, stale-area, and full HashBuild/JoinMap regressions; and git diff --check. All current GitHub CI checks are terminal and green.

Nonblocking repository-state note: the branch is behind current main by 310fa3f50dc, a frontend/parser-only change with no overlapping files or merge-tree conflict. The existing CHANGES_REQUESTED state is from aptend at b95ab95d77 before the c47df29237 fix; aptend or a maintainer still needs to re-review or dismiss that stale review.
